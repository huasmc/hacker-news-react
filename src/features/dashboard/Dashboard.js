import { memo, useEffect, useState } from "react";
import { useDispatch } from "react-redux";
import { Col, Row } from "../../common/styled";
import TabButton from "../../common/tabButton/TabButton";
import { UI_STRINGS } from "../../common/UI_STRINGS";
import NewsFeed from "../newsFeed/NewsFeed";
import { fetchNews } from "./DashboardSlice";
import { DashboardContainer } from "./style";

const Dashboard = () => {
	const [activeTab, setActiveTab] = useState(UI_STRINGS.TAB_BUTTON.ALL);
	const dispatch = useDispatch();

	useEffect(() => {
		dispatch(fetchNews({ query: "angular", page: 0 }));
	}, []);

	return (
		<DashboardContainer>
			<Row>
				<Col span={12}>
					<div style={{ paddingTop: "60px", paddingBottom: "60px" }}>
						<TabButton
							options={[
								UI_STRINGS.TAB_BUTTON.ALL,
								UI_STRINGS.TAB_BUTTON.MY_FAVES,
							]}
							active={activeTab}
							setActive={setActiveTab}
						/>
					</div>
				</Col>
			</Row>
			<NewsFeed />
		</DashboardContainer>
	);
};

export default memo(Dashboard);
