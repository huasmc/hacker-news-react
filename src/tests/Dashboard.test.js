import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import Dashboard from "../features/dashboard/Dashboard";
import {
	fetchNews,
	fetchFavoriteNews,
} from "../features/dashboard/DashboardSlice";

const mockStore = configureStore([]);

describe("Dashboard component", () => {
	let store;
	let news = [
		{ id: 1, title: "News 1" },
		{ id: 2, title: "News 2" },
		{ id: 3, title: "News 3" },
	];

	beforeEach(() => {
		store = mockStore({
			dashboard: {
				news,
			},
		});
		store.dispatch = jest.fn();
	});

	it("should render correctly", () => {
		const { container } = render(
			<Provider store={store}>
				<Dashboard />
			</Provider>
		);
		expect(container).toMatchSnapshot();
	});

	it("should fetch news when page changes", () => {
		render(
			<Provider store={store}>
				<Dashboard />
			</Provider>
		);
		act(() => {
			store.dispatch(fetchNews({ query: "", page: 0, hitsPerPage: 8 }));
		});
		expect(store.dispatch).toHaveBeenCalledWith(
			fetchNews({ query: "", page: 0, hitsPerPage: 8 })
		);
	});

	it('should fetch favorite news when "My Faves" tab is clicked', () => {
		const { getByText } = render(
			<Provider store={store}>
				<Dashboard />
			</Provider>
		);
		act(() => {
			fireEvent.click(getByText("My Faves"));
		});
		expect(store.dispatch).toHaveBeenCalledWith(fetchFavoriteNews(1));
	});

	it("should change the selected filter in the dropdown", () => {
		const { getByText } = render(
			<Provider store={store}>
				<Dashboard />
			</Provider>
		);
		act(() => {
			fireEvent.click(getByText("News 2"));
		});
		expect(store.dispatch).toHaveBeenCalledWith(
			fetchNews({ query: "news 2", page: 0, hitsPerPage: 8 })
		);
	});

	it("should change the active tab when a tab button is clicked", () => {
		const { getByText } = render(
			<Provider store={store}>
				<Dashboard />
			</Provider>
		);
		act(() => {
			fireEvent.click(getByText("My faves"));
		});
		expect(store.dispatch).toHaveBeenCalledTimes(2);
	});
});
