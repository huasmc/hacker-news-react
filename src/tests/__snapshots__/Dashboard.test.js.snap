// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component should render correctly 1`] = `
<div>
  <div
    class="sc-eDWCr beEzaK"
  >
    <div
      class="sc-bcXHqe ROMsa"
    >
      <div
        class="sc-gswNZR bIUWRF"
        span="12"
      >
        <div
          style="padding-top: 60px; padding-bottom: 60px; justify-content: center; align-items: center; display: flex;"
        >
          <div
            class="sc-dkrFOg eCjdnS"
          >
            <button
              class="sc-hLBbgP hqviza"
            >
              All
            </button>
            <button
              class="sc-hLBbgP hvClHC"
            >
              My faves
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-bcXHqe ROMsa"
    >
      <div
        class="sc-gswNZR djDMq"
      >
        <div
          style="padding: 0px 0px 3vw 10vw;"
        >
          <div
            class="dropdown-container"
          >
            <button
              class="dropdown-toggle"
            >
              <table>
                <tbody>
                  <tr>
                    <td>
                      Select your news
                    </td>
                  </tr>
                </tbody>
              </table>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="sc-bcXHqe ROMsa"
        span="12"
      >
        <div
          class="sc-ipEyDJ kZgDPa"
        >
          <div
            class="sc-gswNZR doXCWo"
            span="6"
          >
            <div
              class="sc-gKPRtg gLazcu"
            >
              <div
                class="news-row"
                data-testid="news-item"
              >
                <div
                  class="sc-iBYQkv fnrceR"
                >
                  <table>
                    <tbody>
                      <tr>
                        <td>
                          <img
                            alt=""
                            src="clock.svg"
                          />
                        </td>
                        <td
                          class="news-timestamp"
                        >
                          <span
                            class="sc-eDvSVe bzfGPh"
                          >
                            NaN hour ago by author
                          </span>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                  <span
                    class="sc-jSUZER jNcCqU"
                  >
                     
                  </span>
                </div>
                <div
                  class="sc-ftTHYK bmMHMZ"
                >
                  <img
                    alt=""
                    class="favorite-button"
                    data-testid="heart-unfilled"
                    src="heart-unfilled.svg"
                  />
                </div>
              </div>
            </div>
            <div
              class="sc-gKPRtg gLazcu"
            >
              <div
                class="news-row"
                data-testid="news-item"
              >
                <div
                  class="sc-iBYQkv fnrceR"
                >
                  <table>
                    <tbody>
                      <tr>
                        <td>
                          <img
                            alt=""
                            src="clock.svg"
                          />
                        </td>
                        <td
                          class="news-timestamp"
                        >
                          <span
                            class="sc-eDvSVe bzfGPh"
                          >
                            NaN hour ago by author
                          </span>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                  <span
                    class="sc-jSUZER jNcCqU"
                  >
                     
                  </span>
                </div>
                <div
                  class="sc-ftTHYK bmMHMZ"
                >
                  <img
                    alt=""
                    class="favorite-button"
                    data-testid="heart-unfilled"
                    src="heart-unfilled.svg"
                  />
                </div>
              </div>
            </div>
            <div
              class="sc-gKPRtg gLazcu"
            >
              <div
                class="news-row"
                data-testid="news-item"
              >
                <div
                  class="sc-iBYQkv fnrceR"
                >
                  <table>
                    <tbody>
                      <tr>
                        <td>
                          <img
                            alt=""
                            src="clock.svg"
                          />
                        </td>
                        <td
                          class="news-timestamp"
                        >
                          <span
                            class="sc-eDvSVe bzfGPh"
                          >
                            NaN hour ago by author
                          </span>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                  <span
                    class="sc-jSUZER jNcCqU"
                  >
                     
                  </span>
                </div>
                <div
                  class="sc-ftTHYK bmMHMZ"
                >
                  <img
                    alt=""
                    class="favorite-button"
                    data-testid="heart-unfilled"
                    src="heart-unfilled.svg"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-csuSiG eYwYDZ"
        >
          <div
            class="sc-gswNZR doXCWo"
            span="6"
          />
        </div>
      </div>
      <div
        class="sc-kDvujY fOXXKh"
      >
        <div
          class="sc-pyfCe fuWAEx"
        >
          <span
            style="font-weight: 700; cursor: pointer;"
          >
            &lt;
          </span>
        </div>
        <div>
          <div
            class="sc-jrcTuL ddRbKB"
          >
            1
          </div>
        </div>
        <div
          class="sc-pyfCe fuWAEx"
        >
          <span
            data-testid="next-page"
            style="font-weight: 700; cursor: pointer;"
          >
            &gt;
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
